src/dynamichmc_inference.jl

```diff
@@ -25,16 +27,28 @@ Base.@kwdef struct DynamicHMCPosterior{TA,TP,TD,TT,TR,TS,TK}
     σ_priors::TS
     "Keyword arguments passed on the the ODE solver `solve`."
     solve_kwargs::TK
+    obsvbls::TO
+    sample_u0::Bool
+    adkind::AD
 end
 
 function (P::DynamicHMCPosterior)(θ)
     @unpack parameters, σ = θ
-    @unpack algorithm, problem, data, t, parameter_priors, σ_priors, solve_kwargs = P
-    prob = remake(problem, u0 = convert.(eltype(parameters), problem.u0), p = parameters)
-    solution = solve(prob, algorithm; solve_kwargs...)
-    any((s.retcode ≠ :Success && s.retcode ≠ :Terminated) for s in solution) && return -Inf
-    log_likelihood = sum(sum(logpdf.(Normal.(0.0, σ), solution(t) .- data[:, i]))
-                         for (i, t) in enumerate(t))
+    @unpack algorithm, problem, data, t, parameter_priors = P
+    @unpack σ_priors, solve_kwargs, obsvbls, sample_u0, adkind = P
+    nu, T = length(problem.u0), eltype(parameters)
+    u0 = convert.(T, sample_u0 ? parameters[1:nu] : problem.u0)
+    p = convert.(T, sample_u0 ? parameters[(nu + 1):end] : parameters)
+    _saveat = t === nothing ? Float64[] : t
+    prob = remake(problem, u0 = Tracker.data.(u0), p = Tracker.data.(p))
+    solution = solve(prob, algorithm; saveat = _saveat, solve_kwargs...)
```

Comment:
this will solve twice if Zygote: you might want to just have it be concrete_solve.

Reply:
Yes. If we just keep `concrete_solve`, we won't be able to access the return code of a solution. So should we modify `concrete_solve` to return `solution` + `retcode`? Or we just infer the failure from the shape of `sol`? If `size(sol, 2) < length(saveat)`, then there's a failure?

Reply:
It seems that `size(sol, 2) < length(_saveat)` requires much less better.